### ISSUE TITLE
Refresh of sort order doesn't happen automatically

### ISSUE THREAD

**Issue by BenMorant** [elementary/files#2179]
### What Happened?

Hi everyone,

[default file system, default partionning scheme]

Using pantheon-files, I'm on ~/Downloads . I just finished downloading a .zip . The icon of the archive ("divers 2023-001.zip") doesn't look like an archive type icon + I still get the associated .part file, yet the download has been completed.
Then I navigate somewhere else and come back to  ~/Downloads, only then the view is refreshed, with correct icon and no more .part file.

I can reproduce it anywhere on my desktop, with any file (when I edit any file, a cached file shows up but it doesn't disappear when I close the file I've edited)

i.e I'm in ~/ , I open .zshrc (so, with Code, by default, but I don't think it's relevant here) . I close .zshrc , I still see several LOCK files in the view but they're don't exist anymore ( `ls -a` doesn't list them)
I change the view in files (i.e I go to Downloads),  then I come back to ~/ , and only there the view has been refreshed and the LOCK files aren't present anymore in the view.

I don't know if it's linked to #36 , that issue dates from june 2017 so I prefered to fill a new issue.

A) a .zip archive

1) after the download is finished

![Capture d'écran du 2023-05-01 08 28 43](https://user-images.githubusercontent.com/52816705/235417943-8542e3b4-b04f-4046-98d3-7db644657dbd.png)

2) when I refresh the view by coming back to the folder

![Capture d'écran du 2023-05-01 08 29 01](https://user-images.githubusercontent.com/52816705/235418093-e670df3b-5a50-42c5-8600-622da3961043.png)

B) opening ~/.zshrc

1) before opening

![Captures d'écran de 2023-05-01 10 32 48](https://user-images.githubusercontent.com/52816705/235429909-de81ef25-8215-4678-8374-e2d31eaa35a2.png)

2) after opening

![Captures d'écran de 2023-05-01 10 35 05](https://user-images.githubusercontent.com/52816705/235429960-d3545128-e77b-4c82-b912-9bc103123e45.png)

3) after closing

![Captures d'écran de 2023-05-01 10 35 21](https://user-images.githubusercontent.com/52816705/235429991-2290fd6a-36be-4614-aedd-d9f74c14a30e.png)

4) after refreshing the view by navigating somewhere else and coming back into the folder

![Captures d'écran de 2023-05-01 10 37 26](https://user-images.githubusercontent.com/52816705/235430061-6b6b4d55-4a33-473c-a44e-89bf8909af52.png)

### Steps to Reproduce

1. Using pantheon-files, go to any folder
2.  Edit a file, so a cache file (LOCK) is created
2.b Or download a file on the web, so a part file (.part) is created
3. Finish editing the file (close your editor) 
3.B Or wait for the download process to finish
4. You'll see (well, I'm seeing :) ) that the cache file/the part file is still present in the view

### Expected Behavior

In pantheon-files, the view should automatically refresh after the action (editing a file, downloading an archive...) is finished

[the attached logs : 1) open ~/.zshrc 2) close ~/.zshrc (all the LOCK files should automatically disappear  from the view but they're still present - they aren't physically present on my hard drive (the ls -a command doesn't show them)

### OS Version

7.x (Horus)

### Software Version

Latest release (I have run all updates)

### Log Output

```shell
** (io.elementary.files:20052): DEBUG: 10:17:16.102: Directory.vala:1167: Dir file:///home/ben/.zcompdump-ben-laptop-5.8.1.lock not in cache
** (io.elementary.files:20052): DEBUG: 10:17:16.102: Directory.vala:894: query info failed, Erreur lors de l’obtention des informations du fichier « /home/ben/.zcompdump-ben-laptop-5.8.1.lock » : Aucun fichier ou dossier de ce type file:///home/ben/.zcompdump-ben-laptop-5.8.1.lock

** (io.elementary.files:20052): CRITICAL **: 10:17:16.102: Directory.vala:914: FILE INFO null

** (io.elementary.files:20052): CRITICAL **: 10:17:16.102: files_file_update: assertion '_tmp0_ != NULL' failed
** (io.elementary.files:20052): DEBUG: 10:17:16.103: Directory.vala:894: query info failed, Erreur lors de l’obtention des informations du fichier « /home/ben/.zcompdump-ben-laptop-5.8.1.lock » : Aucun fichier ou dossier de ce type file:///home/ben/.zcompdump-ben-laptop-5.8.1.lock

** (io.elementary.files:20052): CRITICAL **: 10:17:16.103: files_file_update: assertion '_tmp0_ != NULL' failed
** (io.elementary.files:20052): DEBUG: 10:17:16.190: Directory.vala:1167: Dir file:///home/ben/.zsh_history.LOCK not in cache
** (io.elementary.files:20052): DEBUG: 10:17:16.191: Directory.vala:894: query info failed, Erreur lors de l’obtention des informations du fichier « /home/ben/.zsh_history.LOCK » : Aucun fichier ou dossier de ce type file:///home/ben/.zsh_history.LOCK

** (io.elementary.files:20052): CRITICAL **: 10:17:16.191: Directory.vala:914: FILE INFO null

** (io.elementary.files:20052): CRITICAL **: 10:17:16.191: files_file_update: assertion '_tmp0_ != NULL' failed
** (io.elementary.files:20052): DEBUG: 10:17:16.191: Directory.vala:894: query info failed, Erreur lors de l’obtention des informations du fichier « /home/ben/.zsh_history.LOCK » : Aucun fichier ou dossier de ce type file:///home/ben/.zsh_history.LOCK

** (io.elementary.files:20052): CRITICAL **: 10:17:16.191: files_file_update: assertion '_tmp0_ != NULL' failed
** (io.elementary.code:20101): DEBUG: 10:17:16.794: Document.vala:935: Could not find mount location
(io.elementary.code:20101): dconf-DEBUG: 10:17:48.360: change_fast
(io.elementary.code:20101): dconf-DEBUG: 10:17:48.360: change_fast
(io.elementary.code:20101): dconf-DEBUG: 10:17:48.360: change_fast
(io.elementary.code:20101): dconf-DEBUG: 10:17:48.360: change_notify: /io/elementary/code/saved-state/window-position
(io.elementary.code:20101): dconf-DEBUG: 10:17:48.360: change_fast
(io.elementary.code:20101): dconf-DEBUG: 10:17:48.360: change_fast
** (io.elementary.code:20101): DEBUG: 10:17:48.360: Document.vala:377: Closing ".zshrc"
** (io.elementary.code:20101): DEBUG: 10:17:48.360: Document.vala:873: Backup file deleting: /home/ben/.zshrc~
** (io.elementary.code:20101): DEBUG: 10:17:48.360: Document.vala:876: Backup file doesn't exists: /home/ben/.zshrc~

(io.elementary.code:20101): GLib-GObject-WARNING **: 10:17:48.374: ../../../gobject/gsignal.c:2765: instance '0x56260be6f1b0' has no handler with id '566'
(io.elementary.code:20101): dconf-DEBUG: 10:17:48.380: change_fast
(io.elementary.code:20101): dconf-DEBUG: 10:17:48.382: unwatch_fast: "/io/elementary/code/saved-state/" (active: 2, establishing: 0)
(io.elementary.code:20101): dconf-DEBUG: 10:17:48.384: sync
** (io.elementary.files:20052): DEBUG: 10:17:48.395: Directory.vala:1167: Dir file:///home/ben/.zsh_history.LOCK not in cache
** (io.elementary.files:20052): DEBUG: 10:17:48.395: Directory.vala:894: query info failed, Erreur lors de l’obtention des informations du fichier « /home/ben/.zsh_history.LOCK » : Aucun fichier ou dossier de ce type file:///home/ben/.zsh_history.LOCK

** (io.elementary.files:20052): CRITICAL **: 10:17:48.396: Directory.vala:914: FILE INFO null

** (io.elementary.files:20052): CRITICAL **: 10:17:48.396: files_file_update: assertion '_tmp0_ != NULL' failed
** (io.elementary.files:20052): DEBUG: 10:17:48.396: Directory.vala:894: query info failed, Erreur lors de l’obtention des informations du fichier « /home/ben/.zsh_history.LOCK » : Aucun fichier ou dossier de ce type file:///home/ben/.zsh_history.LOCK

** (io.elementary.files:20052): CRITICAL **: 10:17:48.396: files_file_update: assertion '_tmp0_ != NULL' failed
```

### Hardware Info

Dell XPS 13 L322X

**Comment by jeremypw** [comment]
Using the latest version of Files on Horus, I cannot reproduce this issue by following the steps described.  I created the temporary file by editing a text file with Code.  On closing the tab that was being edited the temporary file immediately disappeared from the Files view.

**Comment by jeremypw** [comment]
Ah OK, I can kind of reproduce it using a larger text file and especially if I leave trailing spaces to be autoremoved. In this case there are `.goutputstream` files remaining that do not disappear until the view is refreshed.  However the backup file that remains `<filename>~` is not removed by refreshing so is really being left behind by Code - which is a bug in that app.

**Comment by jeremypw** [comment]
On closer examination it looks like when trailing spaces are being autoremoved, the backup file *is* removed but then a bit later a new one is created - which is a Code issue.  So the confirmed issue here is that `.goutputstream` type files are not properly removed when Files is notified of their deletion.  It is possible that the same bug causes `.lock` files to remain as well.

**Comment by jeremypw** [comment]
Hmm, now I've downloaded and installed the latest master version I cannot reproduce the non-refreshing `.goutputstream` file issue so this should be fixed with the next release.

**Comment by BenMorant** [comment]
Hi @jeremypw , thank you for having investigated !
Have a great day !

**Comment by Gabriel-p** [comment]
I am experiencing this issue all the time. Here's a very simple and reproducible example:

1. Folder with two files

![Captura de pantalla de 2023-07-19 09 30 23](https://github.com/elementary/files/assets/6054951/ad18a327-ff4a-4357-9953-fe9e7e46cc64)

2. Create new file that should be placed at the end because of its name. It is placed first and will not move

![Captura de pantalla de 2023-07-19 09 30 34](https://github.com/elementary/files/assets/6054951/e5febbe6-a691-4644-b032-b9e5a31987d4)

3. Only after refreshing it is placed correctly

![Captura de pantalla de 2023-07-19 09 30 40](https://github.com/elementary/files/assets/6054951/cc107bad-70fa-4a53-b8ca-6d2ca7491720)

This non-refreshing issue is particularly noticeable in Git folders. I'm also constantly having to refresh to see the local thumbnails (these are three images)

![Captura de pantalla de 2023-07-19 09 33 55](https://github.com/elementary/files/assets/6054951/dfceb273-ec23-4bdf-8ea7-27017a4b8c76)

**Comment by jeremypw** [comment]
Yes, there does seem to be a regression in relation to refreshing the order after renaming.  Note that a quick way to refresh the view is to press `F5` or the `Refresh` icon.

**Comment by jeremypw** [comment]
Maybe another regression caused by #2184 not fixed by subsequent commits? If I can fix this without interfering with the ongoing Gtk4 port I will.

**Comment by jeremypw** [comment]
The fix for this in Gtk3 is not simple and I can confirm that the Gtk4 version does not suffer this issue (the views and underlying models are completely different).  Its to do with the Gtk.TreeStore not automatically resorting when new rows are inserted. I'll see if I can find a simpler fix.

**Comment by Gabriel-p** [comment]
The sorting is the least of the issues really. Inside git folders I have to constantly manually refresh (F5) to update files that changed, were removed or created. It has become a hassle sadly

**Comment by Gabriel-p** [comment]
I use Sublime Text projects to manage my workflow. I have several scripts that produce .csv files and/or images that need manual inspection often. But even leaving those types of files aside, whenever I need to delete a script or create a new one I'm confronted with the almost total lack of auto update inside git folders. This is highly inconvenient.

I also use Latex which produces lots of temporary files that remain visible after being removed by the compiler, until I manually update the folder.

**Comment by jeremypw** [comment]
OK, thanks - in that case Code is not suitable as that is focused on code files. I'll see what can be done to fix this in the near future.

**Comment by jeremypw** [comment]
I've fixed (I hope) resorting the view whenever a file is created or renamed.  I'll look at the other issues.

**Comment by jeremypw** [comment]
@Gabriel-p  I am not seeing the temporary files issue with the latest version of Files using Code temporary files as a test case.  It might be better to produce separate issues for each remaining example of non-refreshing (file deletion, thumbnail updating, etc) with a way of reproducing so that I can address them separately - they may have different causes.

**Comment by Gabriel-p** [comment]
Here's an example of what I mean in Git folders. I just removed the cassif.png file and its status icon changed as seen:

![Captura de pantalla de 2023-07-20 09 06 33](https://github.com/elementary/files/assets/6054951/62a4c20a-f2d3-4607-ab39-be8236457582)

The file only goes away after I refresh the folder

**Comment by jeremypw** [comment]
@Gabriel-p When you say "removed" do you mean deleted from disk or removed from the git repository?

**Comment by jeremypw** [comment]
There do seem to be some bugs in the git plugin - the emblems are not updated as expected while a backup file exists (i.e. the document is open in Code) even if the view is refreshed.  But yes, the view is not reflecting emblem changes anyway until refreshed which is related to the thumbnail/icon not updating until refreshed and not related to sorting.  I'll open separate issues I think.

**Comment by jeremypw** [comment]
Hmm, I cannot confirm the issue of file items not disappearing when deleted from disk (or transferred to Trash) using Files but I'll check the code to see if there is any possible way this can occur.

**Comment by Gabriel-p** [comment]
> Hmm, I cannot confirm the issue of file items not disappearing when deleted from disk (or transferred to Trash) using Files but I'll check the code to see if there is any possible way this can occur.

This happens within git folders, not everywhere

**Comment by jeremypw** [comment]
> This happens within git folders, not everywhere

I am not seeing any problems deleting from git folders with Files (with nothing else running).  Is there perhaps some other process using the files delaying their deletion?

Anyway, hopefully the PRs I have pushed today will fix this too. Please feel free to compile and test them if you can.
There are separate PRs fixing sorting, icon update and git status update.

**Comment by OssNass** [comment]
I have a similar problem with latest version of Elementary OS, I use Btrfs file system.

When I rename a folder or a file, the renaming happens, but the file application will not accept it until I refresh, meaning if say I have a "new folder" renamed to "some folder", the new name will appear but when I want to open that "some folder" the file manager tells me that "new folder" doesn't exist, as a result I need to manually refresh the folder in order to accept the new "some folder"

**Comment by OssNass** [comment]
I have several NTFS partition and the problem exists in them as well, Files is not refreshing automatically upon file system changes

**Comment by janxkoci** [comment]
Just FYI I can still observe most of the issues discussed here. Refreshing view in Files is now very deeply embeded in my muscle memory, I do it after almost every operation.

I work almost exclusively inside Dropbox, with some directories also being git projects. I often copy files from SSH servers (sometimes using Files, sometimes with `mc` in the terminal), or create new files from terminal, office apps, sublime text, image editors, etc. I can still reproduce the problem with downloaded files in the first comment on top of this thread.

Caching seems to be severely broken in Files for a while now. :confused:

**Comment by janxkoci** [comment]
My case is just local - I don't expect SSH views to stay live for long, so I don't wait for refresh to be needed (as Files may [freeze](https://github.com/elementary/files/issues/1858) by then). I only open the connections just to copy files already generated. But if I do this from terminal (with `mc` etc) and copy them to my local folder, I need to refresh the local view. Same with files created from apps like office suites, or downloaded from the internet. I mentioned Dropbox and git as git was also mentioned in this thread, so it may be relevant. I use the Dropbox from flathub - there is no UI, icons, anything, even the CLI is broken and needs to be patched to do at least the basic things, but I don't care - all I need is sync between my machines and autostart, which works. If I ever need to share a link, I can use my phone app, the patched CLI, or web.

Some of the issues mentioned are already open, I may open new ones when I have some time.

**Comment by jeremypw** [comment]
I understand you are using v 6.4.1?  You need version 6.5.0 to get the fix for refreshing the view when new files are added.  This issue was a regression caused by changes to make loading views faster.

**Comment by Gabriel-p** [comment]
> Just FYI I can still observe most of the issues discussed here. Refreshing view in Files is now very deeply embeded in my muscle memory, I do it after almost every operation.

I'm at this point too. When will the new version be available?

**Comment by janxkoci** [comment]
Ok I updated today (as we discussed in another thread) and I now did `touch test` in terminal in a folder in one of my Dropbox projects. The file immediatelly showed up in Files too. I will test other issues later, but this looks promising. :slightly_smiling_face:

**Comment by jeremypw** [comment]
@Gabriel-p Version 6.5.0 is already released on Horus (elementary 7) but will not be available on older versions of elementary.
